File: csc/2016/kve/src/main/SynchronizedMap.java
package main;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class SynchronizedMap<K, V> {
    private final Map<K, V> myMap = new HashMap<>();
    private final Object monitor = new Object();

    public V get(K key) {
        synchronized (monitor) {
            return myMap.get(key);
        }
    }

    public void put(K key, V value) {
        synchronized (monitor) {
            myMap.put(key, value);
        }
    }

    public void remove(K key) {
        synchronized (monitor) {
            myMap.remove(key);
        }
    }

    public Set<K> getKeys() {
        synchronized (monitor) {
            return myMap.keySet();
        }
    }

    public Collection<V> getValues() {
        synchronized (monitor) {
            return myMap.values();
        }
    }
}

File: csc/2016/kve/src/collections/SynchronizedMap.kt
package collections

import java.util.*

class SynchronizedMap<K, V> {
    private val myMap = HashMap<K, V>()
    private val monitor = Object()

    @Synchronized operator fun get(key: K): V? {
        return myMap[key]
    }

    @Synchronized fun put(key: K, value: V) {
        myMap.put(key, value)
    }

    fun remove(key: K) {
        synchronized (monitor) {
            myMap.remove(key)
        }
    }

    val keys: Set<K>
        @Synchronized get() = myMap.keys

    val values: Collection<V>
        @Synchronized get() = myMap.values
}
